### PR TITLE
Validate interface implementation

### PR DESCRIPTION
**Pull request by keirlawson**
Validates that an object correctly implements any specified interfaces per #68.

Does not include checking that an interface correctly implements another, nor that an object implements all transitive interfaces.

Feedback appreciated :)

### CODE REVIEW

File: modules/core/src/main/scala/schema.scala
Comment (tpolecat) [inline modules/core/src/main/scala/schema.scala:958]: (The type arguments here and below are unnecessary.)

File: modules/core/src/main/scala/schema.scala
Comment (tpolecat) [inline modules/core/src/main/scala/schema.scala:979]: Should these be compared with `=:=`?

**Review by tpolecat — COMMENTED** [review]
This looks good, thanks! I'll wait for @milessabin to give the final approval.

File: modules/core/src/main/scala/schema.scala
Comment (keirlawson) [reply modules/core/src/main/scala/schema.scala:979]: Do you mean `=!=`?

Comment (keirlawson) [reply modules/core/src/main/scala/schema.scala:979]: So none of these types have an `Eq` instance at present so far as I can see? Happy to implement one if you reckon this is the way to go

Comment (tpolecat) [reply modules/core/src/main/scala/schema.scala:979]: Sorry, `Type` specifically has an `=:=` operator that's like `==` but compares dealiases the types before comparing them. I don't know if it matters here. @milessabin will.

Comment (milessabin) [reply modules/core/src/main/scala/schema.scala:979]: Yes, you should compare types with `=:=` in the typical case.

**Comment by milessabin** [comment]
@keirlawson I'm _really_ sorry it's taken me so long to get to this ... things were a bit hectic in the run up to the end of the year.

Aside from the `==` vs. `=:=` issue this look great :-)

File: modules/core/src/main/scala/schema.scala
Comment (keirlawson) [reply modules/core/src/main/scala/schema.scala:979]: These are `Ast.Type`s though, not `Schema.Type`s, so there doesn't seem to be `=:=` defined unless I'm missing something?

Comment (milessabin) [reply modules/core/src/main/scala/schema.scala:979]: You're absolutely right! Yes, this is correct just as it is.

**Review by milessabin — APPROVED** [review]
LGTM! Many thanks!

**Pull request merged**
